pyrsistent/_pmap.py

```diff
@@ -125,7 +125,27 @@ def __len__(self):
     def __repr__(self):
         return 'pmap({0})'.format(str(dict(self)))
 
-    __eq__ = Mapping.__eq__
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if not isinstance(other, Mapping):
+            return NotImplemented
+        if len(self) != len(other):
+            return False
+        if isinstance(other, PMap):
+            try:
+                if hash(self) != hash(other):
+                    return False
+            except TypeError:
+                # self or other contains non-hashable values
+                pass
+            if self._buckets == other._buckets:
+                return True
+            return dict(self.iteritems()) == dict(other.iteritems())
```

Comment:
In the case you describe, I think we'd have to traverse two, not three times, because a different number of buckets would mean O(1) for checking list equality, because the lengths of the lists differ.
I think, the worst case would rather be two equal pmaps with the same number of buckets, but elements inserted into the same bucket in a different order.
Not sure what to do about that, but I'm going to add a test and benchmark.

The second traversal of the pmap could be eliminated by creating the hashes' underlying frozensets here, and comparing them if their hashes match.

Another problem is that pmaps with unhashable values are tried to be hashed every time - maybe somehow mark the pmap as not hashable in order to skip subsequent hashing attempts? But then again, if we have mutable values in a pmap, they might become hashable later on...